rmf_traffic/test/unit/schedule/test_Mirror.cpp

```diff
@@ -487,3 +488,304 @@ SCENARIO("Testing specialized mirrors")
+    _mirror.update(_database->changes(_all, _mirror.latest_version()));
+  }
+
+  void delay(
+    ParticipantId participant,
+    Duration delay,
+    ItineraryVersion version) final
+  {
+    if (drop_packets)
+      return;
+
+    _database->delay(participant, delay, version);
+    _mirror.update(_database->changes(_all, _mirror.latest_version()));
+  }
+
+  void erase(ParticipantId participant, ItineraryVersion version) final
+  {
+    if (drop_packets)
+      return;
+
+    _database->erase(participant, version);
+    _mirror.update(_database->changes(_all, _mirror.latest_version()));
+  }
+
+  void erase(
+    ParticipantId participant,
+    const std::vector<RouteId>& routes,
+    ItineraryVersion version) final
+  {
+    if (drop_packets)
+      return;
+
+    _database->erase(participant, routes, version);
+    _mirror.update(_database->changes(_all, _mirror.latest_version()));
+  }
+
+  Registration register_participant(ParticipantDescription description) final
+  {
+    // We assume participant registration is done over a reliable connection
+    auto registration = _database->register_participant(std::move(description));
+    _update_participants();
+
+    return registration;
+  }
+
+  void unregister_participant(ParticipantId participant) final
+  {
+    _database->set_current_time(std::chrono::steady_clock::now());
+    _database->unregister_participant(participant);
+    _update_participants();
+  }
+
+  void update_description(
+    ParticipantId participant,
+    ParticipantDescription desc) final
+  {
+    _database->update_description(participant, std::move(desc));
+    _update_participants();
+  }
+
+  void failover()
+  {
+    _database =
+      std::make_shared<rmf_traffic::schedule::Database>(_mirror.fork());
+    _rectifiers->change_database(_database);
+  }
+
+  void rectify()
+  {
+    _rectifiers->rectify();
+  }
+
+  rmf_traffic::schedule::Participant make_participant(
+    rmf_traffic::schedule::ParticipantDescription description)
+  {
+    return rmf_traffic::schedule::make_participant(
+      std::move(description),
+      shared_from_this(),
+      _rectifiers);
+  }
+
+  const rmf_traffic::schedule::Database& database() const
+  {
+    return *_database;
+  }
+
+private:
+
+  void _update_participants()
+  {
+    rmf_traffic::schedule::ParticipantDescriptionsMap descriptions;
+    for (const auto& p : _database->participant_ids())
+      descriptions.insert_or_assign(p, *_database->get_participant(p));
+
+    _mirror.update_participants_info(descriptions);
+  }
+
+  std::shared_ptr<rmf_traffic::schedule::Database> _database;
+  rmf_traffic::schedule::Mirror _mirror;
+  std::shared_ptr<rmf_traffic::schedule::DatabaseRectificationRequesterFactory>
+  _rectifiers;
+
+  rmf_traffic::schedule::Query _all = rmf_traffic::schedule::query_all();
+
+};
+
+//==============================================================================
+SCENARIO("Test forking off of mirrors")
```

Comment:
I know it's only tests, but these could do with a few comments to break up and explain the specific things being tested.

Reply:
Honestly I'm just have it do a bunch of arbitrary operations on the participant, some of which get dropped and then rectified later, all while the database fails over several times, and I just want to see that after each rectification the Database information always matches the participant information, no matter how or when the failovers occur.

I guess I can at least add an explanation to that effect in the test.

Reply:
Updated: 93a96e3